ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/LedgerDB/API.hs

```diff
@@ -762,12 +769,72 @@ instance
   Supporting On-Disk backing stores
 -------------------------------------------------------------------------------}
 
-type LedgerSupportsOnDiskLedgerDB blk =
-  (IndexedMemPack (LedgerState blk EmptyMK) (TxOut (LedgerState blk)))
+type LedgerSupportsLMDBLedgerDB l =
+  (IndexedMemPack (l EmptyMK) (TxOut l), MemPackIdx l EmptyMK ~ l EmptyMK)
 
-type LedgerSupportsLedgerDB blk =
-  ( LedgerSupportsOnDiskLedgerDB blk
-  , LedgerSupportsInMemoryLedgerDB blk
+type LedgerSupportsV2LedgerDB l =
+  (LedgerSupportsInMemoryLedgerDB l, LedgerSupportsLSMLedgerDB l)
+
+class
+  ( LSM.SerialiseKey (TxIn l)
+  , LSM.SerialiseValue (LSMTxOut l)
+  , LSM.ResolveValue (LSMTxOut l)
+  ) =>
+  LedgerSupportsLSMLedgerDB l
+  where
+  -- | LSM trees need to be able to serialize and deserialize values several
+  -- times, without any context. Therefore the approach of using a ledger state to
+  -- hint the era in which values need to be deserialized cannot work with LSM.
+  --
+  -- Therefore, we will instead store 'LSMTxOut' in the LSM database, which will
+  -- be 'TxOut' for most of the unitary blocks and basic hard fork blocks, but
+  -- will be 'ByteArray's for the Cardano Block.
+  --
+  -- In short, the serialisation API for `lsm-tree` looks like this:
+  --
+  -- > data Table k v = ...
+  -- > class SerialiseKey k where
+  -- >   serialiseKey :: k -> ByteArray
+  -- >   deserialiseKey :: ByteArray -> k
+  -- > class SerialiseValue k where
+  -- >   serialiseKey :: v -> ByteArray
+  -- >   deserialiseKey :: ByteArray -> v
+  --
+  -- So we cannot pass a deserialisation hint like we do with
+  -- 'IndexedMemPack', so this option must be discarded:
+  --
+  -- > type Option1 = Table TxIn TxOut
+  -- > instance SerialiseValue TxOut where
+  -- >   deserialiseValue = {- we want to use mempack, but we have no deserialisation hint -}
+  --
+  -- Therefore, we decided to use ByteArrays for Cardano which we later interpret
+  -- with a deserialisation hint in `fromLSMTxOut`:
+  --
+  -- > type Option2 = Table TxIn ByteArray
+  -- > instance SerialiseValue TxOut where
+  -- >   deserialiseValue = id
+  -- > fromLSMTxOut ::  Hint -> ByteArray -> TxOut
+  -- > fromLSMTxOut hint byteArray = {- use mempack with hint on byte array -}
+  -- >
+  -- > -- Roughly what I mean by "consensus deserialises manually outside of lsm-tree"
+  -- > readTxOuts hint txIns table = fromLSMTxOut hint <$> lookups table txIns
+  --
+  -- So we will define `LSMTxOut (LedgerState (CardanoBlock StandardCrypto))` to
+  -- `ByteArray`.
+  type LSMTxOut l
+
+  toLSMTxOut :: Proxy l -> TxOut l -> LSMTxOut l
+  fromLSMTxOut :: l EmptyMK -> LSMTxOut l -> TxOut l
+  lsmIndex :: Proxy l -> LSM.FencePointerIndexType
```

Comment:
What are possible reasons to use different `FencePointerIndexType`s for different blocks? Is it "use `CompactIndex` if possible, otherwise use `OrdinaryIndex`"? Are there blocks that don't support `CompactIndex`?

Reply:
@jorisdral probably has more insights. In general I think blocks whose txins are hashes should use the CompactIndex, and blocks whose txins are something else like consecutive ints (like test blocks) should not use the CompactIndex.

Reply:
Due to the complications we face with the CompactIndex, we are for now going to integrate using the OrdinaryIndex and maybe reconsider doing this in a follow-up PR.